### DISCUSSION TITLE
export as jpeg2000?

### DISCUSSION THREAD

**Discussion by jrochkind** [libvips/libvips#3421] [Q&A] [answered]
The [release notes for 8.11](https://www.libvips.org/2021/06/04/What's-new-in-8.11.html) suggest Jpeg2000 support is now in vips, hooray!

They give a command-line `vips copy` example for writing a TIFF using jp2, but not an example of writing a straight bare jp2 file. 

I have vips 8.14.2 installed. Should this work?

    vips copy some.tiff out.jp2

I'm not sure if this means I don't have jpeg2000 support compiled in, or just that I'm doing it wrong. Is there a way to convert from one format to a .jp2 file with vips?

(Actually, when I try `vips copy in.tiff out.tif[compression=jp2k]`, I get `target output: Compression scheme 33004 scanline encoding is not implemented` -- so maybe I don't have jpeg2000 support compiled in, _and also_ my original command wasn't right anyway?)

**Comment by jcupitt** [comment, marked as answer]
Hi @jrochkind,

Yes, it should work. I expect your libvips is old, or has been built without jp2k support (openjpeg-dev on ubuntu):

I see:

```
$ vips copy k2.jpg x.jp2
$ file x.jp2
x.jp2: JPEG 2000 Part 1 (JP2)
$ 
```

**Reply by jcupitt** [reply]
This should also work:

```
$ vips copy k2.jpg x.tif[compression=jp2k,tile]
$ tiffinfo x.tif
=== TIFF directory 0 ===
TIFF Directory at offset 0x107a38 (1079864)
  Image Width: 1450 Image Length: 2048
  Tile Width: 128 Tile Length: 128
  Resolution: 72.009, 72.009 pixels/inch
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: 33004 (0x80ec)
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
$ 
```

**Reply by jrochkind** [reply]
OK, thanks!  Knowing what is supposed to work is an important part of getting it to work!

My vips at 8.14.2 is not that old. 

I think I installed it on MacOS with `brew install vips`; perhaps the brew package does not successfully build with jp2k support. Gives me something to look into anyway!

**Reply by jrochkind** [reply]
Huh, the brew package has openjpeg as a dependency, which makes it look like it _thinks_ it is building with openjpeg support. https://formulae.brew.sh/formula/vips

I wonder what is going on.

**Reply by jcupitt** [reply]
That's weird. My mac died so I can't test it myself :( 

What does otool say? On linux it's:

```
$ ldd $(which vips) | grep -i openjp2
	libopenjp2.so.7 => /lib/x86_64-linux-gnu/libopenjp2.so.7 (0x00007f20a14cb000)
```

**Reply by jrochkind** [reply]
I am not very experienced in compiling C, so I'm not sure how to run the `otool` command.  it apparently isn't just `otool $(which vips)`, alas. 

    $ otool /opt/homebrew/bin/vips
    error: /Library/Developer/CommandLineTools/usr/bin/otool: one of -fahlLtdoOrTMRIHCGScisPx or --version must be specified

Let's see if I can figure it out...  Aha, `-L`. 

    $ otool /opt/homebrew/bin/vips -L
    /opt/homebrew/bin/vips:
	/opt/homebrew/Cellar/vips/8.14.2/lib/libvips.42.dylib (compatibility version 59.0.0, current version 59.2.0)
	/opt/homebrew/opt/glib/lib/libglib-2.0.0.dylib (compatibility version 7601.0.0, current version 7601.0.0)
	/opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 12.0.0, current version 12.0.0)
	/opt/homebrew/opt/glib/lib/libgobject-2.0.0.dylib (compatibility version 7601.0.0, current version 7601.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)

There's no openjpeg listed in there, even though it is listed as a brew dependency, and is indeed present on my system. 

```
$ brew info openjpeg
==> openjpeg: stable 2.5.0 (bottled), HEAD
Library for JPEG-2000 image manipulation
https://www.openjpeg.org/
/opt/homebrew/Cellar/openjpeg/2.5.0 (533 files, 13.9MB) *
  Poured from bottle on 2022-06-06 at 13:34:23
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openjpeg.rb
License: BSD-2-Clause
```

Does that seem wrong, is that something to point out on an Issue I might file on the brew formula, if I can figure out a way to file such an issue?

**Reply by jcupitt** [reply]
Huh that's odd. You could also try:

```
$ vips --vips-config | grep 2000
JPEG2000 load/save with OpenJPEG: true
```

That'll show if libvips detected openjpeg and tried to compile it in.

I'd open an issue here: https://github.com/Homebrew/homebrew-core/issues

Unfortunately I can't help directly without a mac :(

**Reply by jrochkind** [reply]
It DOES claim to have openjpeg support. Maybe it does a static build instead of linking to the dynamic library?  I only barely know what I'm saying, I'm really not experienced in unix builds. 

~~~
% vips --vips-config | grep 2000
JPEG2000 load/save with OpenJPEG: true
~~~

You provide such an amazing community service with open source vips, which is an incredible package, and your incredibly attentive maintenance and support (some of us notice!) -- it seems like the community ought to be able to buy you a Mac (especially cause it's self-serving of us!), but I am not sure how to make that happen either! (I'm also just an academic/non-profit worker!)

**Reply by jcupitt** [reply]
libvips.org has quite a bit of money from donations. Perhaps I should get a mac from that.

**Comment by jcupitt** [comment]
> (Actually, when I try vips copy in.tiff out.tif[compression=jp2k], I get target output: Compression scheme 33004 scanline encoding is not implemented -- so maybe I don't have jpeg2000 support compiled in, and also my original command wasn't right anyway?)

You need:

```
vips copy in.tiff out.tif[compression=jp2k,tile]
```

You can't write untiled jp2k tiff.

**Reply by jrochkind** [reply]
Ah, I see!  I didn't realize you couldn't write untiled jp2k tiff. The release notes for 8.11 had this example:

    vips copy k2.jpg x.tif[compression=jp2k,lossless]

Is that example wrong, something that won't work?  Do you always need a `tile` in there, actually? 

But I was just trying that to try to debug whether jp2k worked at all in my copy of vips, it isn't actually my use case. My actual use case is straight `.jp2`, not embedded in a TIFF -- _and_ not with tiling or multi-resolution, just a single copy of the image with jp2 compression.  

That should definitely work with `vips copy some.tiff out.jp2`?  

I see the error message I got when I tried that didn't make it into my initial post, apparently I forgot to copy and paste. Let me go look for it and re-report it here.

**Reply by jcupitt** [reply]
Ooop, I've fixed the example, sorry about that. Yes, it should be `vips copy k2.jpg x.tif[compression=jp2k,lossless,tile]`.

Yes, `vips copy k2.jpg x.jp2` ought to work. 

Perhaps the openjpeg on brew is too recent and it's broken libvips somehow? Do you know what version you have? I'm on 2.5.0-1.

**Reply by jrochkind** [reply]
OK, coming back to this:

So I _can_ apparently write with eg: `vips copy k2.jpg x.tif[compression=jp2k,lossless,tile]` 

At least it does this without reporting an error , and produces a non-zero-size file. I'm not sure I have any software that can _read_ a jp2k TIFF, I'm not familiar with that format variation, so I haven't been able to verify it beyond that. 

BUT eg

    vips copy 2y60cl2.tiff out.jp2

HOLD THE PRESSES.  This time that worked (and produced a jp2 that I _do_ have software that can read, and it looks okay). 

OK, since apparently I forgot to copy and paste the error message before, hard to say what was going on or what may have changed, but it seems to be working today. 

Thank you for kind help, sorry for the misdirection! 

I would not be shocked if this issue comes back again, since I still haven't figured out the conditions in which it occurred the first time... we will see!

**Reply by jcupitt** [reply]
Maybe you had a couple of libvipses installed somehow, and it was picking up the wrong one before? Anyway, I'm glad it's working now.
